firmware/Core/Src/adcs_drivers/adcs_internal_drivers.c

```diff
@@ -0,0 +1,203 @@
+				ack_status = ADCS_CMD_Ack(&ack); // confirm telecommand validity by checking the TC Error flag of the last read TC Acknowledge Telemetry Format.
+				if (ack_status != 0 && ack_status != 4) {
+					return ack_status; // there was an error in the command not related to checksum
+				}
+				num_ack_tries++;
+				if (num_ack_tries > ADCS_PROCESSED_TIMEOUT) {
+					return 5; // command failed to process in time
+					// note: sending another telecommand when the first has not been processed
+					// will result in an error in the next telecommand sent
+					// as the ADCS telecommand buffer has been overrun/corrupted
+				}
+			}
+		} else {
+			return cmd_status; // if the HAL had an error, tell us what that is
+		}
+		num_checksum_tries++;
+    } while (ack.error_flag == ADCS_ERROR_FLAG_CRC && num_checksum_tries < ADCS_CHECKSUM_TIMEOUT);  // if the checksum doesn't check out, keep resending the request
+
+	return ack.error_flag; // if the HAL was successful and the ADCS command had an error, tell us what it is
+}
+
+/// @brief Sends a telemetry request over I2C to the ADCS, and resends repeatedly if the checksums don't match.
+/// @param[in] id Valid ADCS telemetry request ID (see Firmware Reference Manual)
+/// @param[in] include_checksum Tells the ADCS whether to use a CRC8 checksum; should be either ADCS_INCLUDE_CHECKSUM or ADCS_NO_CHECKSUM 
+/// @param[in] data_length Length of the data array.
+/// @param[out] data Data array to write the raw telemetry bytes to; length must be at least data_length (should contain the correct number of bytes for the given telemetry request ID)
+/// @return 0 if successful, other numbers if the HAL failed to transmit or receive data. 
+uint8_t ADCS_I2C_telemetry_wrapper(uint8_t id, uint8_t* data, uint32_t data_length, uint8_t include_checksum) {
+	uint8_t checksum_check = ADCS_send_I2C_telemetry_request(id, data, data_length, include_checksum);
+	uint8_t num_checksum_tries = 0;
+	while (checksum_check == 4 && num_checksum_tries < ADCS_CHECKSUM_TIMEOUT) {
+		// if the checksum doesn't check out, keep resending the request up to timeout
+		checksum_check = ADCS_send_I2C_telemetry_request(id, data, data_length, include_checksum);
+		num_checksum_tries++;
+	}
+	return checksum_check;
+}
+
+
+/// @brief Sends a telecommand over I2C to the ADCS.
+/// @param[in] id Valid ADCS telecommand ID (see Firmware Reference Manual)
+/// @param[in] data Data array to send the raw data bytes; length must be at least data_length (should contain the correct number of bytes for the given telecommand ID)
+/// @param[in] data_length Length of the data array.
+/// @param[in] include_checksum Tells the ADCS whether to use a CRC8 checksum; should be either ADCS_INCLUDE_CHECKSUM or ADCS_NO_CHECKSUM 
+/// @return 0 if successful, 4 if the checksums don't match, other numbers if the HAL failed to transmit or receive data.
+uint8_t ADCS_send_I2C_telecommand(uint8_t id, uint8_t* data, uint32_t data_length, uint8_t include_checksum) {
```

Comment:
i2c should be lowercase in function name

Reply:
This is done.